dbms/src/Common/ConcurrentIOQueue.h

```diff
@@ -0,0 +1,219 @@
+                return kickRemaingsOnceWithoutLock(data) ? MPMCQueueResult::OK : res;
+            case MPMCQueueResult::OK:
+                kickRemaingsWithoutLock();
+            default:
+                return res;
+            }
+        }
+        case MPMCQueueResult::OK:
+            kickRemaings();
+        default:
+            return res;
+        }
+    }
+
+    size_t size() const
+    {
+        size_t mpmc_queue_size = mpmc_queue.size();
+        std::lock_guard lock(mu);
+        return mpmc_queue_size + remaings.size();
+    }
+
+    // When the queue is finished, it may appear that the mpmc_queue is empty but the remaings are not, causing isFull to return true.
+    // But this is expected, and the return value of isFull is meaningless after finished/cancelled.
+    bool isFull() const
+    {
+        {
+            std::lock_guard lock(mu);
+            if (!remaings.empty())
+                return true;
+        }
+        return mpmc_queue.isFull();
+    }
+
+    MPMCQueueStatus getStatus() const
+    {
+        return mpmc_queue.getStatus();
+    }
+
+    /// Cancel a NORMAL queue will wake up all blocking readers and writers.
+    /// After `cancel()` the queue can't be pushed or popped any more.
+    /// That means some objects may leave at the queue without poped.
+    bool cancel()
+    {
+        return cancelWith("");
+    }
+    bool cancelWith(String reason)
+    {
+        if (mpmc_queue.cancelWith(std::move(reason)))
+        {
+            std::lock_guard lock(mu);
+            remaings.clear();
+            return true;
+        }
+        return false;
+    }
+
+    const String & getCancelReason() const
+    {
+        return mpmc_queue.getCancelReason();
+    }
+
+    /// Finish a NORMAL queue will wake up all blocking readers and writers.
+    /// After `finish()` the queue can't be pushed any more while `pop` is allowed
+    /// the queue is empty.
+    /// Return true if the previous status is NORMAL.
+    bool finish()
+    {
+        return mpmc_queue.finish();
+    }
+
+private:
+    void kickRemaingsWithoutLock()
+    {
+        while (!remaings.empty() && mpmc_queue.tryPush(std::move(remaings.back())) == MPMCQueueResult::OK)
+            remaings.pop_back();
+    }
+
+    void kickRemaings()
+    {
+        std::lock_guard lock(mu);
+        kickRemaingsWithoutLock();
+    }
+
+    bool kickRemaingsOnceWithoutLock(T & data)
+    {
+        if (!remaings.empty())
+        {
+            data = std::move(remaings.back());
+            remaings.pop_back();
+            return true;
+        }
+        return false;
+    }
+
+    bool kickRemaingsOnce(T & data)
+    {
+        std::lock_guard lock(mu);
+        return kickRemaingsOnceWithoutLock(data);
+    }
+
+private:
+    mutable std::mutex mu;
+    // Used to hold objs pushed by `NonBlockingPush` that exceed the capacity.
+    // remaings will try to push into the mpmc_queue when pop/tryPop.
+    std::deque<T> remaings;
```

Comment:
`remaings` is not a word. Should be `remains` or `remaining_objs`

Reply:
ok, renamed to `remaining_objs`